src/core/api/scan.go

```diff
@@ -0,0 +1,199 @@
+	repoName := sa.GetString(":splat")
+	tag := sa.GetString(":tag")
+	projectName, _ := utils.ParseRepository(repoName)
+
+	pro, err := sa.ProjectMgr.Get(projectName)
+	if err != nil {
+		sa.SendInternalServerError(errors.Wrap(err, "scan API: prepare"))
+		return
+	}
+	if pro == nil {
+		sa.SendNotFoundError(errors.Errorf("project %s not found", projectName))
+		return
+	}
+	sa.pro = pro
+
+	// Check authentication
+	if !sa.SecurityCtx.IsAuthenticated() {
+		sa.SendUnAuthorizedError(errors.New("Unauthorized"))
+		return
+	}
+
+	// Assemble artifact object
+	digest, err := digestFunc(repoName, tag, sa.SecurityCtx.GetUsername())
+	if err != nil {
+		sa.SendInternalServerError(errors.Wrap(err, "scan API: prepare"))
+		return
+	}
+
+	sa.artifact = &v1.Artifact{
+		NamespaceID: pro.ProjectID,
+		Repository:  repoName,
+		Tag:         tag,
+		Digest:      digest,
+		MimeType:    v1.MimeTypeDockerArtifact,
+	}
+
+	logger.Debugf("Scan API receives artifact: %#v", sa.artifact)
+}
+
+// Scan artifact
+func (sa *ScanAPI) Scan() {
+	// Check access permissions
+	resource := rbac.NewProjectNamespace(sa.pro.ProjectID).Resource(rbac.ResourceScan)
+	if !sa.SecurityCtx.Can(rbac.ActionCreate, resource) {
+		sa.SendForbiddenError(errors.New(sa.SecurityCtx.GetUsername()))
+		return
+	}
+
+	if err := scan.DefaultController.Scan(sa.artifact); err != nil {
+		sa.SendInternalServerError(errors.Wrap(err, "scan API: scan"))
+		return
+	}
+
+	sa.Ctx.ResponseWriter.WriteHeader(http.StatusAccepted)
+}
+
+// Report returns the required reports with the given mime types.
+func (sa *ScanAPI) Report() {
+	// Check access permissions
+	resource := rbac.NewProjectNamespace(sa.pro.ProjectID).Resource(rbac.ResourceScan)
+	if !sa.SecurityCtx.Can(rbac.ActionRead, resource) {
+		sa.SendForbiddenError(errors.New(sa.SecurityCtx.GetUsername()))
+		return
+	}
+
+	// Extract mime types
+	producesMimes := make([]string, 0)
+	if hl, ok := sa.Ctx.Request.Header[v1.HTTPAcceptHeader]; ok && len(hl) > 0 {
+		producesMimes = append(producesMimes, hl...)
+	}
+
+	// Get the reports
+	reports, err := scan.DefaultController.GetReport(sa.artifact, producesMimes)
+	if err != nil {
+		sa.SendInternalServerError(errors.Wrap(err, "scan API: get report"))
+		return
+	}
+
+	vulItems := make(map[string]interface{})
+	for _, rp := range reports {
+		// Resolve scan report data only when it is ready
+		if len(rp.Report) == 0 {
+			continue
+		}
+
+		vrp, err := report.ResolveData(rp.MimeType, []byte(rp.Report))
+		if err != nil {
+			sa.SendInternalServerError(errors.Wrap(err, "scan API: get report"))
+			return
+		}
+
+		vulItems[rp.MimeType] = vrp
+	}
+
+	sa.Data["json"] = vulItems
+	sa.ServeJSON()
+}
+
+// Log returns the log stream
+func (sa *ScanAPI) Log() {
+	// Check access permissions
+	resource := rbac.NewProjectNamespace(sa.pro.ProjectID).Resource(rbac.ResourceScan)
+	if !sa.SecurityCtx.Can(rbac.ActionRead, resource) {
+		sa.SendForbiddenError(errors.New(sa.SecurityCtx.GetUsername()))
+		return
+	}
```

Comment:
There is a method RequireProjectAccess in BaseController, that can do permission checking simpler.